docs/user/managing-devices.md

```diff
@@ -505,6 +507,99 @@ LABEL appType="compose"
 
 ## Using Device Lifecycle Hooks
 
+You can use device lifecycle hooks to make the agent run user-defined commands at specific points in the device's lifecycle. For example, you can add a shell script to your OS images that backs up your application data and then specify that this script shall be run and complete successfully before the agent can start updating the system.
+
+The following device lifecycle hooks are supported:
+
+| Lifecycle Hook | Description |
+| -------------- | ----------- |
+| `beforeUpdating` | This hook is called after the agent completed preparing for the update and before actually making changes to the system. If an action in this hook returns with failure, the agent aborts the update. |
+| `afterUpdating` | This hook is called after the agent has written the update to disk. If an action in this hook returns with failure,the agent will abort and roll back the update. |
+| `beforeRebooting` | This hook is called before the system reboots. The agent will block the reboot until running the action has completed. If any action in this hook returns with failure, the agent will abort and roll back the update. |
+| `afterRebooting` | This hook is called when the agent first starts after a reboot. If any action in this hook returns with failure, the agent will report this but continue starting up. |
```

Comment:
what if we rollback the OS? should it fire then or not.

Reply:
According to the state diagram in #637 it should not. What would least surprise users / be most useful to users, though?

Reply:
I think it would be easier for the user not to reason about what `kind` of reboot scenario as long as there is a reboot. but it is tricky